src/autorest.bicep/src/type-generator.ts

```diff
@@ -463,6 +463,12 @@ export function generateTypes(host: AutorestExtensionHost, definition: ProviderD
       flags |= ObjectTypePropertyFlags.WriteOnly;
     }
 
+    // There is ambiguity between OpenApi v2 & v3 about whether properties can be readonly & required.
+    // We can work around this by just stripping out required if a property is readonly.
+    if (flags & ObjectTypePropertyFlags.ReadOnly) {
+      flags &= ~ObjectTypePropertyFlags.Required;
+    }
```

Comment:
Should we be handling this in Bicep instead? The type system there knows whether we're dealing with an input vs output, and the `Required` flag on a property in the output does signal that the property will always be present.

Reply:
Initially I was against this because of the confusion in the OpenAPI spec - v2 and v3 conflict with each other, whereas the type system doesn't differentiate. My plan was to align with v2, but I think you're right - it makes more sense to align the type system representation to v3. Will investigate the changes in Bicep.

v2:
<img width="697" height="218" alt="image" src="https://github.com/user-attachments/assets/6f4a9b0a-07d0-47b3-9a6a-a44a527c8068" />
v3:
<img width="693" height="260" alt="image" src="https://github.com/user-attachments/assets/4085897e-9f6e-4e5b-8ca1-da469e02c6f0" />

Reply:
PR to fix this in Bicep: https://github.com/Azure/bicep/pull/17913. I'll close this one.